File: homework4/PathFinder.java
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Scanner;
import java.util.Set;
import java.io.File;

public class PathFinder {
    public static final String COMMA_DELIMITER = ",";

    static class Node {
        int id;
        String name;
        double lat, lon;
        double d = Double.MAX_VALUE; // initial distance, infinite
        double f = Double.MAX_VALUE; // estimate
        Integer predecessor = null;

        Node(int id, String name, double lat, double lon) {
            this.id = id;
            this.name = name;
            this.lat = lat;
            this.lon = lon;
        }
    }

    static class Edge {
        int target;
        double weight; 

        Edge(int target, double weight) {
            this.target = target;
            this.weight = weight;
        }
    }

    Map<Integer, List<Edge>> graph = new HashMap<>();
    Map<Integer, Node> nodes = new HashMap<>();

    public PathFinder(String nodesFile, String edgesFile) throws FileNotFoundException {
        loadNodes(nodesFile);
        loadEdges(edgesFile);
    }

    private void loadEdges(String filename) throws FileNotFoundException {
        try (Scanner scanner = new Scanner(new File(filename))) {
            if (scanner.hasNextLine()) scanner.nextLine(); // header
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine().trim();
                if (line.isEmpty()) continue;
                String[] values = line.split(COMMA_DELIMITER);
                int src = Integer.parseInt(values[0].trim());
                int target = Integer.parseInt(values[1].trim());

                double distance = haversine(nodes.get(src), nodes.get(target));
                
                if (!graph.containsKey(src)) {
                    graph.put(src, new ArrayList<>());
                }
                graph.get(src).add(new Edge(target, distance));
                if (!graph.containsKey(target)) {
                    graph.put(target, new ArrayList<>());
                }
                graph.get(target).add(new Edge(src, distance));
            }
        }
    }

    private void loadNodes(String filename) throws FileNotFoundException {
        try (Scanner scanner = new Scanner(new File(filename))) {
            if (scanner.hasNextLine()) scanner.nextLine(); // header
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine().trim();
                if (line.isEmpty()) continue;
                String[] values = line.split(COMMA_DELIMITER);
                int stopId = Integer.parseInt(values[0].trim());
                nodes.put(stopId, new Node(
                    stopId, values[1].trim(), Double.parseDouble(values[2].trim()), Double.parseDouble(values[3].trim())
                ));
            }
        }
    }

    private double haversine(Node a, Node b) {
        // convert to radians 
        double lat = Math.toRadians(a.lat);
        double lat2 = Math.toRadians(b.lat);
        double lonDiff = Math.toRadians(b.lon - a.lon);
        
        //inside square root
        double beforeSquare = (Math.pow(Math.sin((lat - lat2) / 2), 2)) + (Math.cos(lat)* Math.cos(lat2) * (Math.pow(Math.sin((lonDiff) / 2), 2)));
        return 6371 * 2 * Math.asin(Math.sqrt(beforeSquare));
    }

    private void initAlgo(int startId, int endId) {
        for (Node v : nodes.values()) {
            v.d = Double.MAX_VALUE;
            v.f = Double.MAX_VALUE;
            v.predecessor = null;
        }
        Node s = nodes.get(startId);
        s.d = 0;
        s.f = haversine(s, nodes.get(endId));
    }

    private void relaxAlgo(Node u, Node v, double weight, int endId, PriorityQueue<Node> Q) {
        if (v.d > u.d + weight) {
            v.d = u.d + weight;
            v.f = v.d + haversine(v, nodes.get(endId));
            v.predecessor = u.id;

            Q.remove(v);
            Q.add(v);
        }
    }

    public List<Integer> algo(int startId, int endId) {

        if (!nodes.containsKey(startId) || !nodes.containsKey(endId)) {
            return null;
        }
        
        if (!graph.containsKey(startId)) {
            return null;
        }

        initAlgo(startId, endId);

        PriorityQueue<Node> Q = new PriorityQueue<>((a, b) -> {
            int comp = Double.compare(a.f, b.f);
            if (comp != 0) return comp;
            return Integer.toString(a.id).compareTo(Integer.toString(b.id));
        });

        Q.addAll(nodes.values());

        Set<Integer> S = new HashSet<>();

        while (!Q.isEmpty()) {
            Node u = Q.poll();

            if (u.id == endId) {
                return reconstructPath(endId);
            }

            S.add(u.id);

            for (Edge edge : graph.getOrDefault(u.id, Collections.emptyList())) {
                Node v = nodes.get(edge.target);
                if (S.contains(v.id)) continue;
                relaxAlgo(u, v, edge.weight, endId, Q);
            }
        }
        return null; // no path
    }

    private List<Integer> reconstructPath(int endId) {
        LinkedList<Integer> path = new LinkedList<>();
        Integer current = endId;

        while (current != null) {
            path.addFirst(current);
            current = nodes.get(current).predecessor;
        }
        return path;
    }

    public static void main(String[] args) {
        try {
            if (args.length < 5) {
                System.out.println("Not enough arguments");
                return;
            }
    
            String nodesFile = args[0];
            String edgesFile = args[1];
            int source = Integer.parseInt(args[3]);
            int goal = Integer.parseInt(args[4]);
    
            PathFinder pf = new PathFinder(nodesFile, edgesFile);
            List<Integer> path = pf.algo(source, goal);
    
            if (path == null) {
                System.out.println("NONE");
            } else {
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < path.size(); i++) {
                    if (i > 0) sb.append("->");
                    sb.append(path.get(i));
                }
                System.out.println(sb);
    
                double total = 0;
                for (int i = 0; i < path.size() - 1; i++) {
                    total += pf.haversine(pf.nodes.get(path.get(i)), pf.nodes.get(path.get(i + 1)));
                }
                System.out.printf("%.2f%n", total);
            }
        } catch (FileNotFoundException e) {
            System.out.println("File not found: " + e.getMessage());
        }
    }
}

File: homework4/pathfinder.py
import csv
import math
import heapq
import sys


class Node:
    def __init__(self, id, name, lat, lon):
        self.id = id
        self.name = name
        self.lat = lat
        self.lon = lon
        self.d = float('inf')   
        self.f = float('inf')   
        self.predecessor = None

class PathFinder:
    def __init__(self, nodes_file, edges_file):
        self.nodes = {}        
        self.graph = {}   
        self.load_nodes(nodes_file)
        self.load_edges(edges_file)

    def load_nodes(self, filename):
        with open(filename, newline='') as f:
            reader = csv.reader(f)
            next(reader)  # skip header
            for row in reader:
                stop_id = int(row[0].strip())
                self.nodes[stop_id] = Node(
                    stop_id,
                    row[1].strip(),
                    float(row[2].strip()),
                    float(row[3].strip())
                )

    def load_edges(self, filename):
        with open(filename, newline='') as f:
            reader = csv.reader(f)
            next(reader)  # skip header
            for row in reader:
                src = int(row[0].strip())
                tgt = int(row[1].strip())
                dist = self.haversine(self.nodes[src], self.nodes[tgt])

                
                if src not in self.graph:
                    self.graph[src] = []
                self.graph[src].append((tgt, dist))

                if tgt not in self.graph:
                    self.graph[tgt] = []
                self.graph[tgt].append((src, dist))

    def haversine(self, a, b):
        lat1 = math.radians(a.lat)
        lat2 = math.radians(b.lat)
        d_lat = math.radians(b.lat - a.lat)
        d_lon = math.radians(b.lon - a.lon)

        h = math.sin(d_lat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(d_lon / 2) ** 2
        return 6371 * 2 * math.asin(math.sqrt(h))

    def init_algo(self, start_id, end_id):
        for v in self.nodes.values():
            v.d = float('inf')
            v.f = float('inf')
            v.predecessor = None
        s = self.nodes[start_id]
        s.d = 0
        s.f = self.haversine(s, self.nodes[end_id])

    def relax_algo(self, u, v, weight, end_id, heap, counter):
        if v.d > u.d + weight:
            v.d = u.d + weight
            v.f = v.d + self.haversine(v, self.nodes[end_id])
            v.predecessor = u.id
            
            heapq.heappush(heap, (v.f, str(v.id), next(counter), v))

    def algo(self, start_id, end_id):
        self.init_algo(start_id, end_id)

        counter = iter(range(10**9))  # unique tiebreaker to avoid comparing Nodes

        # Build initial heap from all nodes
        heap = []
        for node in self.nodes.values():
            heapq.heappush(heap, (node.f, str(node.id), next(counter), node))

        visited = set()  # HashSet<Integer>

        while heap:
            f, _, __, u = heapq.heappop(heap)

            if u.id in visited:
                continue 

            if u.id == end_id:
                return self.reconstruct_path(end_id)

            visited.add(u.id)

            for (tgt_id, weight) in self.graph.get(u.id, []):
                v = self.nodes[tgt_id]
                if v.id in visited:
                    continue
                self.relax_algo(u, v, weight, end_id, heap, counter)

        return None  # no path found

    def reconstruct_path(self, end_id):
        path = []
        current = end_id
        while current is not None:
            path.append(current)
            current = self.nodes[current].predecessor
        path.reverse()
        return path

if __name__ == "__main__":
    if len(sys.argv) < 6:
        print("Not enough arguments")
        sys.exit(1)

    nodes_file = sys.argv[1]
    edges_file = sys.argv[2]
    
    source_id = int(sys.argv[4])
    target_id = int(sys.argv[5])

    pf = PathFinder(nodes_file, edges_file)
    path = pf.algo(source_id, target_id)

    if path is None:
        print("NONE")
    else:
        
        print("->".join(str(n) for n in path))

        total = sum(
            pf.haversine(pf.nodes[path[i]], pf.nodes[path[i+1]])
            for i in range(len(path) - 1)
        )
        print(f"{total:.2f}")